packages/wonder-blocks-core/components/with-ssr-placeholder.js

```diff
@@ -72,49 +85,74 @@ export default class WithSSRPlaceholder extends React.Component<Props, State> {
-        // we are reliably told a WithSSRPlaceholder component further
-        // up the chain already handled our SSR case.
-        if (alreadyPerformedFirstRender) {
-            // We need to stop the forced second render and to do that, we have
-            // to influence our componentDidMount. Fortunately, that occurs
-            // after this, so let's save a member value to then use there.
-            this._alreadyPerformedFirstRender = alreadyPerformedFirstRender;
-            return children();
-        }
+        // We are the first component in the tree.
+        // We are in control of instigating a second render for our
+        // component tree.
+        this._isTheRootComponent = true;
 
-        // We are mounted and we no higher SSR component handled the first
-        // render so let's tell our children that we did.
         if (mounted) {
+            // This is our second non-SSR render, so let's tell everyone to
+            // do their thing.
             return (
-                <HasHadFirstRenderContext.Provider value={true}>
+                <RenderStateContext.Provider value="standard">
                     {children()}
-                </HasHadFirstRenderContext.Provider>
+                </RenderStateContext.Provider>
             );
         }
 
-        // We don't have a parent that handled SSR, and this must be our
-        // first render, and we have a placeholder, so render the placeholder.
+        // OK, this is the very first render.
+        // If we have a placeholder, we render it, and ensure that any
+        // nested SSR components know we're still on that first render
+        // but they're not in charge of instigating the second render.
         if (placeholder) {
-            return placeholder();
+            return (
+                <RenderStateContext.Provider value="initial">
+                    {placeholder()}
+                </RenderStateContext.Provider>
+            );
         }
 
-        // We don't have a parent that handled SSR, and this must be our
-        // first render and we don't have a placeholder, so return nothing.
+        // Otherwise, we return nothing.
         return null;
     }
 
+    _maybeRender(renderState: RenderState) {
+        const {children, placeholder} = this.props;
+
+        switch (renderState) {
+            case "root":
+                return this._renderAsRootComponent();
+
+            case "initial":
+                // We're not the root component, so we just have to either
+                // render our placeholder or nothing.
+                // The second render is going to be triggered for us.
+                if (placeholder) {
+                    return placeholder();
+                }
+                // Otherwise, we render nothing.
+                return null;
+
+            case "standard":
+                // We have covered the SSR render, we're now rendering normally.
```

Comment:
I wonder if `"normal"` is a better constant for this case.

Reply:
I considered `normal` but I think it's an ambiguous term these days (what is normal?); I felt `standard` had less baggage.

Reply:
In that case maybe the comment should be update to match the constant.